docs/decisions/infra/0010-feature-flags-system-design.md

```diff
@@ -0,0 +1,85 @@
+
+## Feature management
+
+One key design question is how features should be managed once defined. How should a team go about enabling and disabling the feature flags and adjusting the percentage of traffic to send to a new feature during a feature launch?
+
+### Option 1. Manage features using app-config module as part of service layer
+
+Define features in [app-config](/infra/app/app-config/), and use that configuration in the [service layer](/infra/app/service/) to create and configure the features in AWS Evidently.
+
+* Everything is defined in code and in one place.
+* Feature and feature configurations are updated automatically as part of service deploys or can be done manually via a terraform apply.
+
+The configuration in the app-config module might look something like the following:
+
+```terraform
+features = {
+  some_disabled_feature = {}
+
+  some_enabled_feature = {
+    enabled_variation = 1
+  }
+
+  simple_ab_test = {
+    throttle_percentages = [0.5]
+  }
+
+  test_with_multiple_variations = {
+    throttle_percentages = [0.33, 0.33, 0.33]
+  }
+}
+```
+
+### Option 2. Manage features using app-config as part of a separate infrastructure layer
+
+Define features in [app-config](/infra/app/app-config/main.tf). Create the features in the [service layer](/infra/app/service/) but set things like throttle percentages (for gradual rollouts) in a separate infrastructure layer.
+
+* Allows for separation of permissions. For example, individuals can have permission to update feature launch throttle percentages without having permission to create, edit, or delete the features themselves.
+
+### Option 3. Manage features in AWS Console outside of terraform
+
+Define features in [app-config](/infra/app/app-config/main.tf) and create them in the [service layer](/infra/app/service), but set things like throttle percentages (for gradual rollouts) outside of terraform (e.g. via AWS Console). Use `lifecycle { ignore_changes = [entity_overrides] }` in the terraform configuration for the `aws_evidently_feature` resources to ignore settings that are managed via AWS Console.
+
+* Empowers non-technical roles like business owners and product managers to set enable and disable feature flags and adjust feature launch throttle percentages without needing to depend on the development team.
+* A no-code approach using the AWS Console GUI means that it's possible to leverage the full set of functionality offered by AWS CloudWatch Evidently, including things like scheduled launches, with minimal training and without needing to learn how to do it in code.
+
+A reduced configuration in the app-config module that just defines the features might look something like the following:
+
+```terraform
+feature_flags = [
+  "some_new_feature_1", "some_new_feature_2"
+]
+
+experiments = [
```

Comment:
I'm lacking familiarity with Evidently so not totally sure what these attributes are, but in general I like the idea of defining the feature flags as code since that avoids needing to manually add a flag to each environment in the AWS console, and I like the idea of using the AWS Console for more advanced configuration of those flags when the configuration is environment-specific.

Reply:
Oh oops, I actually didn't meant to include the "experiments" / "A/B tests" idea in here. While working on this tech spec I think adding support for A/B testing is more complex and can be deferred to later. I think just having feature flags i.e. "disabled"/"enabled" is all we need. I'll remove this block.